src/backend/distributed/commands/schema_based_sharding.c

```diff
@@ -227,18 +569,150 @@ citus_internal_unregister_tenant_schema_globally(PG_FUNCTION_ARGS)
 	if (HeapTupleIsValid(namespaceTuple))
 	{
 		ReleaseSysCache(namespaceTuple);
-
 		ereport(ERROR, (errmsg("schema is expected to be already dropped "
 							   "because this function is only expected to "
 							   "be called from Citus drop hook")));
 	}
-
 	uint32 tenantSchemaColocationId = SchemaIdGetTenantColocationId(schemaId);
 
 	DeleteTenantSchemaLocally(schemaId);
 	SendCommandToWorkersWithMetadata(TenantSchemaDeleteCommand(schemaNameStr));
 
 	DeleteColocationGroup(tenantSchemaColocationId);
+	PG_RETURN_VOID();
+}
+
+
+/*
+ * citus_schema_distribute gets a regular schema name, then converts it to a tenant
+ * schema.
+ */
+Datum
+citus_schema_distribute(PG_FUNCTION_ARGS)
+{
+	CheckCitusVersion(ERROR);
+
+	Oid schemaId = PG_GETARG_OID(0);
+	char *schemaName = get_namespace_name(schemaId);
+	EnsureTenantSchemaAllowed(schemaId);
+	EnsureSchemaOwner(schemaId);
+	EnsureCoordinator();
+
+	/* Prevent any update to the schema */
+	LockDatabaseObject(NamespaceRelationId, schemaId, 0, AccessExclusiveLock);
```

Comment:
Not actually, left it to make my comment at top to make sense to remember the problem.

Reply:
I couldn't quite follow. Which things get blocked by this? (also, do we need isolation tests?)

Reply:
I mean we should handle if a new table is inserted into the schema while distributing the schema. `LockDatabaseObject` is useless for or purpose but left it so to replace it with proper method.

Isolation tests make sense to show how we expect the other operations are blocked or not.

Reply:
I added isolation tests for schema_distribute/undistribute. Here are my observations.

 

**Concurrent create table and drop schema**
```c
/* Prevent concurrent schema deletion and table creation under the schema */
LockDatabaseObject(NamespaceRelationId, schemaId, 0, AccessExclusiveLock);
```

Above lock prevents concurrent table creation under the schema since create table statements take an `AccessShareLock` on the schema object. If we do not take the lock, a table, which is added concurrently, could be left as a PG local table while all other tables under the schema are converted to single shard during `citus_schema_distribute`. The similar situation for `citus_schema_undistribute`: concurrent table is created as single shard but all other tables are converted back to PG locals. The lock also prevents concurrent schema drop. (schema rename is ok)

**Concurrent drop/alter table** 

Concurrent table drops or modifications are not blocked by the schema lock, but when a table is dropped or altered, its oid is simply non-existent during the `citus_schema_distribute/undistribute`. I do existence check before starting to distribute/undistribute the schema to not have the error after a long time (early error).